lib/Hsm.lua

```diff
@@ -0,0 +1,121 @@
+local Fsm = require(script.Parent.Fsm)
+
+local Hsm = {}
+
+function Hsm.new()
+	local hsm = {}
+	hsm.fsms = {}
+
+	local function hsmEvent(event)
+		local handlingFsmIndex = nil
+		for i = #hsm.fsms, 1, -1 do
+			if hsm.fsms[i][event] then
+				handlingFsmIndex = i
+			end
+		end
+		if handlingFsmIndex == nil then
+			-- There is no such action on any fsm, so return nil.
+			return nil
+		else
+			return function(...)
+				-- Remove nested FSMs we won't need now.
+				for i = #hsm.fsms, handlingFsmIndex + 1, -1 do
+					-- Call the on_leave_ function for the current state, if there is one.
+					local fsm = hsm.fsms[i]
+					if fsm.current and fsm["on_leave_" .. fsm.current] then
+						fsm["on_leave_" .. fsm.current](fsm, event, fsm.current, "none", ...)
+					end
+
+					table.remove(hsm.fsms)
+				end
+
+				-- Call the event function on the FSM that can handle it.
+				local handlingFsm = hsm.fsms[handlingFsmIndex]
+				handlingFsm[event](...)
+			end
+		end
+	end
+
+	setmetatable(
+		hsm,
+		{
+			__index = function(_, key)
+				if Hsm[key] then
```

Comment:
`return Hsm[key] or hsmEvent(key)`?

Reply:
Is it right to be polluting the Hsm namespace with arbitrary name bindings?

Reply:
It is assumed that the names are events, and if there is no such event, it will resolve to nil, so calling the event will cause an error, as for the underlying fsm library.